File: wear/wear/src/main/java/androidx/wear/ongoingactivity/OngoingActivityData.java
/*
 * Copyright 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package androidx.wear.ongoingactivity;

import android.app.Notification;
import android.app.PendingIntent;
import android.graphics.drawable.Icon;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;
import androidx.core.content.LocusIdCompat;
import androidx.versionedparcelable.ParcelField;
import androidx.versionedparcelable.ParcelUtils;
import androidx.versionedparcelable.VersionedParcelable;
import androidx.versionedparcelable.VersionedParcelize;

/**
 * This class is used internally by the library to represent the data of an OngoingActivity.
 */
@VersionedParcelize
public class OngoingActivityData implements VersionedParcelable {
    @Nullable
    @ParcelField(value = 1, defaultValue = "null")
    private final Icon mAnimatedIcon;

    @NonNull
    @ParcelField(value = 2)
    private final Icon mStaticIcon;

    @Nullable
    @ParcelField(value = 3, defaultValue = "null")
    private OngoingActivityStatus mStatus;

    @NonNull
    @ParcelField(value = 4)
    private final PendingIntent mTouchIntent;

    @Nullable
    @ParcelField(value = 5, defaultValue = "null")
    private final LocusIdCompat mLocusId;

    @ParcelField(value = 6, defaultValue = "-1")
    private final int mOngoingActivityId;

    OngoingActivityData(
            @Nullable Icon animatedIcon,
            @NonNull Icon staticIcon,
            @Nullable OngoingActivityStatus status,
            @NonNull PendingIntent touchIntent,
            @Nullable LocusIdCompat locusId,
            int ongoingActivityId
    ) {
        mAnimatedIcon = animatedIcon;
        mStaticIcon = staticIcon;
        mStatus = status;
        mTouchIntent = touchIntent;
        mLocusId = locusId;
        mOngoingActivityId = ongoingActivityId;
    }

    @NonNull
    NotificationCompat.Builder extend(@NonNull NotificationCompat.Builder builder) {
        ParcelUtils.putVersionedParcelable(builder.getExtras(), EXTRA_ONGOING_ACTIVITY,
                this);
        return builder;
    }

    @NonNull Notification extendAndBuild(@NonNull NotificationCompat.Builder builder) {
        Notification notification = extend(builder).build();
        // TODO(http://b/169394642): Undo this if/when the bug is fixed.
        notification.extras.putBundle(
                EXTRA_ONGOING_ACTIVITY,
                builder.getExtras().getBundle(EXTRA_ONGOING_ACTIVITY)
        );
        return notification;
    }

    /**
     * Checks if the given notification represents an ongoing activity.
     */
    public static boolean hasOngoingActivity(@NonNull Notification notification) {
        return notification.extras.getBundle(EXTRA_ONGOING_ACTIVITY) != null;
    }

    /**
     * Deserializes the {@link OngoingActivityData} from a notification.
     *
     * Applies defaults from the notification for information not provided as part of the
     * {@link OngoingActivity}.
     *
     * @param notification the notification that may contain information about a Ongoing
     *                     Activity.
     * @return the data, or null of the notification doesn't contain Ongoing Activity data.
     */
    @Nullable
    public static OngoingActivityData create(@NonNull Notification notification) {
        return ParcelUtils.getVersionedParcelable(notification.extras, EXTRA_ONGOING_ACTIVITY);
    }

    /**
     * Get the animated icon that can be used on some surfaces to represent this
     * {@link OngoingActivity}. For example, in the WatchFace.
     */
    @Nullable
    public Icon getAnimatedIcon() {
        return mAnimatedIcon;
    }

    /**
     * Get the static icon that can be used on some surfaces to represent this
     * {@link OngoingActivity}. For example in the WatchFace in ambient mode.
     */
    @NonNull
    public Icon getStaticIcon() {
        return mStaticIcon;
    }

    /**
     * Get the status of this ongoing activity, the status may be displayed on the UI to
     * show progress of the Ongoing Activity.
     */
    @Nullable
    public OngoingActivityStatus getStatus() {
        return mStatus;
    }

    /**
     * Get the intent to be used to go back to the activity when the user interacts with the
     * Ongoing Activity in other surfaces (for example, taps the Icon on the WatchFace)
     */
    @NonNull
    public PendingIntent getTouchIntent() {
        return mTouchIntent;
    }

    /**
     * Get the LocusId of this {@link OngoingActivity}, this can be used by the launcher to
     * identify the corresponding launcher item and display it accordingly.
     */
    @Nullable
    public LocusIdCompat getLocusId() {
        return mLocusId;
    }

    /**
     * Give the id to this {@link OngoingActivity}, as a way to reference it in
     * [fromExistingOngoingActivity]
     */
    public int getOngoingActivityId() {
        return mOngoingActivityId;
    }

    // Status is mutable, by the library.
    void setStatus(@NonNull OngoingActivityStatus status) {
        mStatus = status;
    }

    /** Notification action extra which contains ongoing activity extensions */
    private static final String EXTRA_ONGOING_ACTIVITY =
            "android.wearable.ongoingactivities.EXTENSIONS";

    static final int DEFAULT_ID = -1;
}

File: wear/wear/src/main/java/androidx/wear/ongoingactivity/OngoingActivityData.kt
package androidx.wear.ongoingactivity

import android.app.Notification
import android.app.PendingIntent
import android.graphics.drawable.Icon
import android.os.Build
import android.os.Bundle
import androidx.annotation.RequiresApi
import androidx.annotation.RestrictTo
import androidx.core.app.NotificationCompat
import androidx.core.content.LocusIdCompat

/**
 * This class is used internally by the library to represent the data of an OngoingActivity.
 */
public class OngoingActivityData {
    /** See [OngoingActivity.Builder.setAnimatedIcon] */
    public var animatedIcon: Icon? = null

    /** See [OngoingActivity.Builder.setStaticIcon] */
    public var staticIcon: Icon? = null

    /** See [OngoingActivity.Builder.setStatus] */
    public var status: OngoingActivityStatus? = null

    /** See [OngoingActivity.Builder.setTouchIntent] */
    public var touchIntent: PendingIntent? = null

    /** See [OngoingActivity.Builder.setLocusId] */
    public var locusId: LocusIdCompat? = null

    /** See [OngoingActivity.Builder.setOngoingActivityId] */
    public var ongoingActivityId: Int = DEFAULT_ID

    public constructor() {
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public fun extend(builder: NotificationCompat.Builder): NotificationCompat.Builder =
        builder.apply {
            val bundle = Bundle()
            animatedIcon?.let { bundle.putParcelable(KEY_ANIMATED_ICON, it) }
            staticIcon?.let { bundle.putParcelable(KEY_STATIC_ICON, it) }
            status?.let { it.extend(bundle) }
            touchIntent?.let { bundle.putParcelable(KEY_TOUCH_INTENT, it) }
            locusId?.let { builder.setLocusId(locusId) }
            if (ongoingActivityId != DEFAULT_ID) {
                bundle.putInt(KEY_ID, ongoingActivityId)
            }
            extras.putBundle(EXTRA_ONGOING_ACTIVITY, bundle)
        }

    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public fun extendAndBuild(builder: NotificationCompat.Builder): Notification {
        return extend(builder).build().apply {
            // TODO(http://b/169394642): Undo this if/when the bug is fixed.
            extras.putBundle(
                EXTRA_ONGOING_ACTIVITY,
                builder.extras.getBundle(EXTRA_ONGOING_ACTIVITY)
            )
        }
    }

    public companion object {
        /**
         * Internal version that deserializes the [OngoingActivityData] from a notification.
         */
        @JvmStatic
        @RestrictTo(RestrictTo.Scope.LIBRARY)
        public fun createInternal(notification: Notification): OngoingActivityData? =
            (notification.extras[EXTRA_ONGOING_ACTIVITY] as? Bundle)?.let { bundle ->
                OngoingActivityData().apply {
                    animatedIcon = bundle[KEY_ANIMATED_ICON] as? Icon
                    staticIcon = bundle[KEY_STATIC_ICON] as? Icon
                    status = OngoingActivityStatus.create(bundle)
                    touchIntent = bundle[KEY_TOUCH_INTENT] as? PendingIntent
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                        locusId = Api29Impl.getLocusId(notification)
                    }
                    ongoingActivityId = bundle.getInt(KEY_ID, DEFAULT_ID)
                }
            }

        /**
         * Deserializes the [OngoingActivityData] from a notification.
         *
         * Applies defaults from the notification for information not provided as part of the
         * [OngoingActivity].
         *
         * @param notification the notification that may contain information about a Ongoing
         * Activity.
         * @return the data, or null of the notification doesn't contain Ongoing Activity data.
         */
        @JvmStatic
        public fun create(notification: Notification): OngoingActivityData? =
            createInternal(notification)?.apply {
                if (animatedIcon == null) {
                    animatedIcon = notification.smallIcon
                }
                if (staticIcon == null) {
                    staticIcon = notification.smallIcon
                }
                if (status == null) {
                    status = notification.extras.getString(Notification.EXTRA_TEXT)?.let {
                        TextOngoingActivityStatus(it)
                    }
                }
                if (touchIntent == null) {
                    touchIntent = notification.contentIntent
                }
            }
    }

    // Inner class required to avoid VFY errors during class init.
    @RequiresApi(29)
    private object Api29Impl {
        @JvmStatic
        fun getLocusId(notification: Notification): LocusIdCompat? =
            notification.locusId?.let { LocusIdCompat.toLocusIdCompat(it) }
    }
}

/** Notification action extra which contains ongoing activity extensions  */
private const val EXTRA_ONGOING_ACTIVITY = "android.wearable.ongoingactivities.EXTENSIONS"

// Keys within EXTRA_ONGOING_ACTIVITY_EXTENDER for ongoing activities options.
private const val KEY_ANIMATED_ICON = "animatedIcon"
private const val KEY_STATIC_ICON = "staticIcon"
private const val KEY_TOUCH_INTENT = "touchIntent"
private const val KEY_ID = "id"

private const val DEFAULT_ID = -1